### PR TITLE
feat(oceanbase-catalog): Support schema operations for OceanBase JDBC catalog.

### PR DESCRIPTION
**Pull request by yuanoOo**
### What changes were proposed in this pull request?

- Support schema operations for OceanBase JDBC catalog.
- Add OceanBaseContainer to test schema operations.

### Why are the changes needed?

Fix: #4990 

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

Add unit test: TestOceanBaseDatabaseOperations.

### CODE REVIEW

**Comment by yuanoOo** [comment]
@mchades PTAL.

File: integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/OceanBaseContainer.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import org.rnorth.ducttape.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.wait.strategy.Wait;
+
+public class OceanBaseContainer extends BaseContainer {
+  public static final Logger LOG = LoggerFactory.getLogger(OceanBaseContainer.class);
+
+  public static final String DEFAULT_IMAGE = "oceanbase/oceanbase-ce:4.2.1-lts";
```
Comment (mchades) [inline integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/OceanBaseContainer.java:45]: Why are you using version 4.2.1? I noticed that the latest version is 4.3.1.

Comment (yuanoOo) [reply integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/OceanBaseContainer.java:45]: The main reason is that it has the LTS suffix, which means it is more stable. So the latest version was not selected.

Comment (mchades) [reply integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/OceanBaseContainer.java:45]: OK. This makes sense

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java [hunk 2]
```diff
@@ -45,17 +55,62 @@ private static Set<String> createSysOceanBaseDatabaseNames() {
   public String generateCreateDatabaseSql(
       String databaseName, String comment, Map<String, String> properties) {
 
-    throw new UnsupportedOperationException("Not implemented yet.");
+    String originComment = StringIdentifier.removeIdFromComment(comment);
```
[the remark this answers is not in the document]
Comment (yuanoOo) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:58]: I checked and found that PG and Doris do not have the logic of `StringIdentifier.addToComment(identifier, comment)`.

I tried refactoring the code for reuse:
```
  /**
   * @param databaseName The name of the database.
   * @param comment The comment of the database.
   * @param properties The properties of the database.
   * @return the SQL statement to create a database with the given name and comment.
   */
  protected String generateCreateDatabaseSql(
      String databaseName, String comment, Map<String, String> properties) {
    String originComment = StringIdentifier.removeIdFromComment(comment);
    return generateCreateDatabaseSql(databaseName, comment, originComment, properties);
  }

  protected String generateCreateDatabaseSql(
          String databaseName, String comment, String originComment, Map<String, String> properties) {
    throw new UnsupportedOperationException("");
  };
```
- In this way, there is no need to change the code of PG and Doris modules.
- Oceanbase and MySQL can choose to override the newly added methods: `protected String generateCreateDatabaseSql(String databaseName, String comment, String originComment, Map<String, String> properties) `

I don't think there is much need. what do you think?

[the remark this answers is not in the document]
Comment (mchades) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:58]: I think you can add such method:
```java
protected abstract boolean supportSchemaComment();
```
then handle the comment before call method `generateCreateDatabaseSql()`, WDYT?

[the remark this answers is not in the document]
Comment (mchades) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:58]: FYI:
```java
@Override
  public void create(String databaseName, String comment, Map<String, String> properties)
      throws SchemaAlreadyExistsException {
    LOG.info("Beginning to create database {}", databaseName);
    String originComment = StringIdentifier.removeIdFromComment(comment);
    if (!supportSchemaComment() && StringUtils.isNotEmpty(originComment)) {
        throw new UnsupportedOperationException(
            "Doesn't support setting schema comment: " + originComment);
    }
    try (final Connection connection = getConnection()) {
      JdbcConnectorUtils.executeUpdate(
          connection, generateCreateDatabaseSql(databaseName, comment, properties));
      LOG.info("Finished creating database {}", databaseName);
    } catch (final SQLException se) {
      throw this.exceptionMapper.toGravitinoException(se);
    }
  }
```
then move the implementation of method `generateCreateDatabaseSql(
      String databaseName, String comment, Map<String, String> properties)` in Mysql to the default implementation.

> Do I need to open a new PR for it?

yeah, make changes in this PR so we can see the impact immediately.

[the remark this answers is not in the document]
Comment (yuanoOo) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java]: Ok, I refactored it.

File: catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java [hunk 3]
```diff
@@ -129,4 +137,13 @@ protected Connection getConnection() throws SQLException {
   protected boolean isSystemDatabase(String dbName) {
     return false;
   }
+
+  /**
+   * Check whether support setting schema comment.
+   *
+   * @return true for all cases.
+   */
+  protected boolean supportSchemaComment() {
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java:146]: This should be abstract

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java [hunk 4]
```diff
@@ -44,22 +52,65 @@ private static Set<String> createSysOceanBaseDatabaseNames() {
   @Override
   public String generateCreateDatabaseSql(
```
Comment (mchades) [inline catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:53]: we can make this method as parent's default implementation for mysql and oceanbase reusing

File: catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java (hunk 3 above)
Comment (yuanoOo) [reply catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java]: Done.

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java (hunk 4 above)
Comment (yuanoOo) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java]: Fixed.

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java [hunk 5]
```diff
@@ -41,25 +47,56 @@ private static Set<String> createSysOceanBaseDatabaseNames() {
...
-  public String generateCreateDatabaseSql(
-      String databaseName, String comment, Map<String, String> properties) {
-
-    throw new UnsupportedOperationException("Not implemented yet.");
-  }
-
   @Override
   public String generateDropDatabaseSql(String databaseName, boolean cascade) {
-    throw new UnsupportedOperationException("Not implemented yet.");
+    final String dropDatabaseSql = String.format("DROP DATABASE `%s`", databaseName);
```
Comment (mchades) [inline catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:52]: Can this method be reused?

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java [hunk 6]
```diff
@@ -41,25 +47,56 @@ private static Set<String> createSysOceanBaseDatabaseNames() {
...
+    } catch (SQLException sqlException) {
+      throw this.exceptionMapper.toGravitinoException(sqlException);
+    }
+    return dropDatabaseSql;
   }
 
   @Override
   public JdbcSchema load(String databaseName) throws NoSuchSchemaException {
-    throw new UnsupportedOperationException("Not implemented yet.");
+    List<String> allDatabases = listDatabases();
```
Comment (mchades) [inline catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java:78]: also this

File: catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java [hunk 7]
```diff
@@ -114,8 +115,15 @@ protected void dropDatabase(String databaseName, boolean cascade) {
    * @param properties The properties of the database.
    * @return the SQL statement to create a database with the given name and comment.
    */
-  protected abstract String generateCreateDatabaseSql(
-      String databaseName, String comment, Map<String, String> properties);
+  protected String generateCreateDatabaseSql(
```
[the remark this answers is not in the document]
Comment (yuanoOo) [reply catalogs/catalog-jdbc-common/src/main/java/org/apache/gravitino/catalog/jdbc/operation/JdbcDatabaseOperations.java:127]: Added.

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java (hunk 5 above)
Comment (yuanoOo) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java]: Fixed. I also refactored `org.apache.gravitino.catalog.jdbc.operation.JdbcDatabaseOperations#isSystemDatabase`

File: catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java (hunk 6 above)
Comment (yuanoOo) [reply catalogs/catalog-jdbc-oceanbase/src/main/java/org/apache/gravitino/catalog/oceanbase/operation/OceanBaseDatabaseOperations.java]: Fixed.

**Comment by jerryshao** [comment]
@mchades can you please move forward this PR?

**Review by mchades — APPROVED** [review]
LGTM. Thanks for your contributions and sorry for my later response.

I think the next PR should be `support simple table operations for Oceanbase catalog`

**Comment by yuanoOo** [comment]
> LGTM. Thanks for your contributions and sorry for my later response.
> 
> I think the next PR should be `support simple table operations for Oceanbase catalog`

I think it might be a better choice to implement the `org.apache.gravitino.catalog.oceanbase.converter` package first. Both schema and table operations depend on it, and we can test it while testing schema and table operations.What do you think?

**Comment by mchades** [comment]
> > LGTM. Thanks for your contributions and sorry for my later response.
> > I think the next PR should be `support simple table operations for Oceanbase catalog`
> 
> I think it might be a better choice to implement the `org.apache.gravitino.catalog.oceanbase.converter` package first. Both schema and table operations depend on it, and we can test it while testing schema and table operations.What do you think?

ok, plz go ahead

**Pull request merged**